runtime/common/src/parachains.rs

```diff
@@ -304,8 +470,53 @@ decl_module! {
 			Ok(())
 		}
 
+		/// Provide a proof that some validator has commited a double-vote.
+		///
+		/// The weight is 0; in order to avoid DoS a `SignedExtension` validation
+		/// is implemented.
+		#[weight = SimpleDispatchInfo::FixedNormal(0)]
+		pub fn report_double_vote(
+			origin,
+			report: DoubleVoteReport<
+				<T::KeyOwnerProofSystem as KeyOwnerProofSystem<(KeyTypeId, Vec<u8>)>>::Proof,
+				T::Hash,
+			>,
+		) -> DispatchResult {
+			let reporter = ensure_signed(origin)?;
+
+			let validators = <session::Module<T>>::validators();
+			let validator_set_count = validators.len() as u32;
+
+			let session_index = report.proof.session();
+
+			// We have already checked this proof in `SignedExtension`, but we need
+			// this here to get the full identification of the offender.
+			let offender = T::KeyOwnerProofSystem::check_proof(
+					(PARACHAIN_KEY_TYPE_ID, report.identity.encode()),
+					report.proof.clone(),
```

Comment:
clone is probably unnecessary here. i'd prefer to avoid the clone as the `FullIdentification` will be heavy